sunpy/coordinates/wcs_utils.py

```diff
@@ -153,22 +153,31 @@ def solar_wcs_frame_mapping(wcs):
     if rsun is not None:
         frame_args['rsun'] = rsun
 
-    # Truncate the ctype to the first four letters
-    ctypes = {c[:4] for c in wcs.wcs.ctype}
+    frame_class = _frame_class_from_ctypes(wcs.wcs.ctype)
 
-    if {'HPLN', 'HPLT'} <= ctypes:
-        return Helioprojective(**frame_args)
+    if frame_class:
+        if frame_class == HeliographicStonyhurst:
+            frame_args.pop('observer', None)
+        if frame_class == Heliocentric:
+            frame_args.pop('rsun', None)
 
-    if {'HGLN', 'HGLT'} <= ctypes:
-        frame_args.pop('observer', None)
-        return HeliographicStonyhurst(**frame_args)
+        return frame_class(**frame_args)
 
-    if {'CRLN', 'CRLT'} <= ctypes:
-        return HeliographicCarrington(**frame_args)
 
-    if {'SOLX', 'SOLY'} <= ctypes:
-        frame_args.pop('rsun', None)
-        return Heliocentric(**frame_args)
+def _frame_class_from_ctypes(ctypes):
+    # Truncate the ctype to the first four letters
+    ctypes = {c[:4] for c in ctypes}
+
+    mapping = {
+        Helioprojective: {'HPLN', 'HPLT'},
+        HeliographicStonyhurst: {'HGLN', 'HGLT'},
+        HeliographicCarrington: {'CRLN', 'CRLT'},
+        Heliocentric: {'SOLX', 'SOLY'},
+    }
+
+    for frame_class, ctype_pair in mapping.items():
+        if ctype_pair <= ctypes:
+            return frame_class
```

Comment:
I need to think about this more, but this limits us to only our frames, where as at the moment there's no reason you couldn't make a map in equatorial coordinates or anything else in astropy, for that reason I am not really in favour of it.

Reply:
Well, this private function is used in two places:

- In `wcs_utils.py`, it merely breaks out the behavior that is currently in `solar_wcs_frame_mapping()`.  That is, `solar_wcs_frame_mapping()` didn't return a frame for such non-SunPy frames, and it still doesn't.
- In `mapbase.py`, it's used to determine whether we should include the observer information in the WCS instance.  What I can do is change this PR so that  `GenericMap.wcs` includes the observer information unless both the `CTYPEi` corresponds to a SunPy frame class and that frame class does not have the `observer` frame attribute.   Thus, the behavior for non-SunPy frames is unchanged.

Reply:
> What I can do is change this PR so that  `GenericMap.wcs` includes the observer information unless both the `CTYPEi` corresponds to a SunPy frame class and that frame class does not have the `observer` frame attribute.   Thus, the behavior for non-SunPy frames is unchanged.

Done